### PR TITLE
NFQ queue range v6

### PR DESCRIPTION
**Pull request by gozzy**
Make sure these boxes are signed before submitting your Pull Request -- thank you.

- [x] I have read the contributing guide lines at https://redmine.openinfosecfoundation.org/projects/suricata/wiki/Contributing
- [x] I have signed the Open Information Security Foundation contribution agreement at https://suricata-ids.org/about/contribution-agreement/
- [x] I have updated the user guide (in doc/userguide/) to reflect the changes made (if applicable)

Describe changes:
- added ability to specify Netfilter queue range with '-q' option (similar to iptables '--queue-balance' option)
- increase maximum number of queues from 16 to 65535
- NFQ contexts are created dynamically instead of on-stack arrays

This replaces #3594.

### CODE REVIEW

File: src/source-nfq.c [hunk 1]
```diff
@@ -804,65 +822,120 @@ TmEcode VerdictNFQThreadDeinit(ThreadVars *tv, void *data)
...
+ *  \retval -1 on failure.
+ */
+int NFQParseAndRegisterQueues(const char *queues)
+{
+    uint16_t queue_start = 0;
+    uint16_t queue_end = 0;
+    uint16_t num_queues = 1;    // if argument is correct, at least one queue will be created
+
+    // Either "id" or "start:end" format (e.g., "12" or "0:5")
+    int count = sscanf(queues, "%hu:%hu", &queue_start, &queue_end);
```
Comment (victorjulien) [inline src/source-nfq.c]: I'm not familiar with this 'h' modifier use in sccanf, what does it do?

**Comment by victorjulien** [comment]
I have no objection to the array approach, but I do think it would be nicer if we could have a much more dynamic solution. Instead of a large array that is indexed by queue num use a list or hash or something. I don't really see use cases for more than a couple of hundred queues max, so this array will be very lightly populated if there is only a single high queue number.

Any ideas?

File: src/source-nfq.c (hunk 1 above)
Comment (gozzy) [reply src/source-nfq.c]: In this case 'h' modifies 'u' and means 'unsigned short int'. Without 'h' it would have been 'unsigned int'.

**Comment by victorjulien** [comment]
On 29-01-19 13:13, Alexander Gozman wrote:
> *@gozzy* commented on this pull request.
> 
> ------------------------------------------------------------------------
> 
> In src/source-nfq.c
> <https://github.com/OISF/suricata/pull/3609#discussion_r251806907>:
> 
>> +/**
> + *  \brief Parses and adds Netfilter queue(s).
> + *
> + *  \param string with the queue number or range
> + *
> + *  \retval 0 on success.
> + *  \retval -1 on failure.
> + */
> +int NFQParseAndRegisterQueues(const char *queues)
> +{
> +    uint16_t queue_start = 0;
> +    uint16_t queue_end = 0;
> +    uint16_t num_queues = 1;    // if argument is correct, at least one queue will be created
> +
> +    // Either "id" or "start:end" format (e.g., "12" or "0:5")
> +    int count = sscanf(queues, "%hu:%hu", &queue_start, &queue_end);
> 
> In this case 'h' modifies 'u' and means 'unsigned short int'. Without
> 'h' it would have been 'unsigned int'.

Oh cool, so is it equivalent to doing %5u or is it even smarter and
takes the max value USHRT_MAX into account?

**Comment by gozzy** [comment]
> Oh cool, so is it equivalent to doing %5u or is it even smarter and takes the max value USHRT_MAX into account?

Some sscanf implementations that I saw used different internal flags for 'u' and 'hu' and, as a result, do different type casts. I suppose that if we want unsigned short we should let sscanf know about it :)

**Comment by gozzy** [comment]
> I have no objection to the array approach, but I do think it would be nicer if we could have a much more dynamic solution. Instead of a large array that is indexed by queue num use a list or hash or something. I don't really see use cases for more than a couple of hundred queues max, so this array will be very lightly populated if there is only a single high queue number.
> 
> Any ideas?

Do I understand correctly that it would be better to turn  *g_nfq_t and *g_nfq_q into lists or anything other than array?

**Comment by gozzy** [comment]
@inliniac, @regit 
Are there any more things to do or current version is acceptable? :)

BTW, I have one more fix on the way: NFQ does not increase packets counter and "iface-stat" command returns zero.

**Comment by victorjulien** [comment]
I think the list would be nice as a future improvement, but is not a requirement for this PR. I will test this PR and merge if I don't find issues.

Wrt the stats, happy to see a separate PR for that. Thanks!

**Comment by gozzy** [comment]
Sure, the stats will be fixed in a separate PR a bit later (there's a dependency on this one).

**Comment by victorjulien** [comment]
Merged in #3641, thanks Alexander!

**Pull request closed without merging**